pkg/ruler/compat.go

```diff
@@ -143,10 +143,17 @@ func EngineQueryFunc(engine *promql.Engine, q storage.Queryable, overrides Rules
 	}
 }
 
-func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter) rules.QueryFunc {
+func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter, queryTime *prometheus.CounterVec, userID string) rules.QueryFunc {
 	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
 		queries.Inc()
 
+		var startTime time.Time
+		// If we've been passed a counter vec we want to record the wall time spent executing this request.
+		if queryTime != nil {
+			startTime = time.Now()
+			defer func() { queryTime.WithLabelValues(userID).Add(float64(time.Since(startTime))) }()
```

Comment:
This is in nanoseconds, should we use that or round to a more granular time duration?

Reply:
Based on the name `cortex_ruler_query_seconds_total`, I'm thinking it should be seconds. 😄